src/users/src/main/java/engineering/everest/lhotse/users/services/DefaultUsersService.java

```diff
@@ -1,45 +1,71 @@
-    private final RandomFieldsGenerator randomFieldsGenerator;
-    private final PasswordEncoder passwordEncoder;
+    private final KeycloakSynchronizationService keycloakSynchronizationService;
 
-    public DefaultUsersService(HazelcastCommandGateway commandGateway,
-                               RandomFieldsGenerator randomFieldsGenerator,
-                               PasswordEncoder passwordEncoder) {
+    public DefaultUsersService(HazelcastCommandGateway commandGateway, KeycloakSynchronizationService keycloakSynchronizationService) {
         this.commandGateway = commandGateway;
-        this.randomFieldsGenerator = randomFieldsGenerator;
-        this.passwordEncoder = passwordEncoder;
+        this.keycloakSynchronizationService = keycloakSynchronizationService;
     }
 
     @Override
-    public void updateUser(UUID requestingUserId, UUID userId, String emailChange,
-                           String displayNameChange, String passwordChange) {
+    public void updateUser(UUID requestingUserId, UUID userId, String emailChange, String displayNameChange) {
+        commandGateway.sendAndWait(new UpdateUserDetailsCommand(userId, emailChange, displayNameChange, requestingUserId));
+    }
 
-        commandGateway.sendAndWait(new UpdateUserDetailsCommand(userId, emailChange,
-                displayNameChange, encodePasswordIfNotBlank(passwordChange), requestingUserId));
+    @Override
+    public void updateUserRoles(UUID requestingUserId, UUID userId, Set<Role> roles) {
+        commandGateway.sendAndWait(new UpdateUserRolesCommand(userId, roles, requestingUserId));
     }
 
     @Override
-    public UUID createUser(UUID requestingUserId, UUID organizationId, String username, String displayName, String rawPassword) {
-        return commandGateway.sendAndWait(new CreateUserCommand(randomFieldsGenerator.genRandomUUID(), organizationId, requestingUserId,
-                username, encodePasswordIfNotBlank(rawPassword), displayName));
+    public UUID createUser(UUID requestingUserId, UUID organizationId, String username, String displayName) {
+        try {
+            // Create user in the keycloak first so that we can get a newly created user id and map it in our app db.
+            keycloakSynchronizationService.createUser(
+                    Map.of("username", username,
+                            "email", username,
+                            "enabled", true,
+                            "attributes", new UserAttribute(organizationId, Set.of(Role.ORG_USER), "Guest"),
+                            "credentials", List.of(
+                                    Map.of("type", "password",
+                                            "value", "changeme",
+                                            "temporary", true))));
+
+            return commandGateway.sendAndWait(new CreateUserCommand(getUserId(username), organizationId,
+                    requestingUserId, username, displayName));
```

Comment:
The reason why we have to do this way is, because we have to establish a relationship between keycloak users and our app users. So that we can have more control over what additional extra fields(name, profile pic) a user can have irrespective of what keycloak provides. For example, if we want to capture and display user hobbies in profile section, we can capture those details using our app api endpoint and store it in projections database.

We have to create a user in keycloak first and then map it in our app because we are not able to override a user id in keycloak. A user id is randomly generated by keycloak irrespective of what we pass. So we have to rely on what it returns.

Reply:
While we could make the core application the source of truth, the motivation behind adopting Keycloak (also, being able to more easily swap it out for another IdP) is that we want to be able to support identity federation.

Keycloak can be configured to allow multiple IdPs so the use case for authenticating using your Google, GitHub, social media or corporate account.

Authentication is a supporting domain and not a concern for the core domain.

With this approach, all the core domain cares about is that you have been granted a verifiable access token.

We then need to bootstrap the user and their default organisation. The contrived example domain is stretching a bit thin here - we may end up moving to project oriented model where you can invite others to join you.